File: packages/protocol/dart/lib/src/device_signing.dart
import 'dart:convert';
import 'dart:typed_data';

const deviceScopes = <String>[
  'observe',
  'send',
  'interrupt',
  'approve',
  'administer',
];

final _payloadMagic = utf8.encode('agent-talk-signed-payload\u0000v1');
final _domainPattern = RegExp(r'^agent-talk\/[a-z0-9.-]+\/v1$');
final _fieldNamePattern = RegExp(r'^[a-z][a-z0-9_.-]*$');
final _sha256Pattern = RegExp(r'^[0-9a-f]{64}$');
const _maximumPayloadBytes = 1024 * 1024;

enum DeviceSigningErrorCode {
  invalidDomain,
  invalidField,
  duplicateField,
  payloadTooLarge,
  invalidScope,
  duplicateScope,
  invalidSha256,
}

class DeviceSigningContractException implements Exception {
  const DeviceSigningContractException(this.code, this.message);

  final DeviceSigningErrorCode code;
  final String message;

  @override
  String toString() => 'DeviceSigningContractException: $message';
}

class SignedPayloadField {
  const SignedPayloadField(this.name, this.value);

  final String name;
  final Object value;
}

Uint8List _uint32(int value) {
  final bytes = ByteData(4)..setUint32(0, value, Endian.big);
  return bytes.buffer.asUint8List();
}

Uint8List _bytes(Object value) {
  if (value is String) {
    return Uint8List.fromList(utf8.encode(value));
  }
  if (value is List<int>) {
    if (value.any((byte) => byte < 0 || byte > 255)) {
      throw const DeviceSigningContractException(
        DeviceSigningErrorCode.invalidField,
        'A signed payload byte value is out of range.',
      );
    }
    return Uint8List.fromList(value);
  }
  throw const DeviceSigningContractException(
    DeviceSigningErrorCode.invalidField,
    'A signed payload field must contain UTF-8 text or bytes.',
  );
}

void _addFrame(BytesBuilder output, List<int> value) {
  if (value.length > _maximumPayloadBytes) {
    throw const DeviceSigningContractException(
      DeviceSigningErrorCode.payloadTooLarge,
      'A signed payload field exceeds one MiB.',
    );
  }
  output
    ..add(_uint32(value.length))
    ..add(value);
}

Uint8List canonicalSignedPayload(
  String domain,
  List<SignedPayloadField> fields,
) {
  if (!_domainPattern.hasMatch(domain)) {
    throw const DeviceSigningContractException(
      DeviceSigningErrorCode.invalidDomain,
      'The device-signature domain is invalid.',
    );
  }
  final names = <String>{};
  final output = BytesBuilder(copy: false)..add(_payloadMagic);
  _addFrame(output, utf8.encode(domain));
  output.add(_uint32(fields.length));
  for (final field in fields) {
    if (!_fieldNamePattern.hasMatch(field.name)) {
      throw const DeviceSigningContractException(
        DeviceSigningErrorCode.invalidField,
        'A device-signature field name is invalid.',
      );
    }
    if (!names.add(field.name)) {
      throw const DeviceSigningContractException(
        DeviceSigningErrorCode.duplicateField,
        'A device-signature field name is duplicated.',
      );
    }
    _addFrame(output, utf8.encode(field.name));
    _addFrame(output, _bytes(field.value));
  }
  final payload = output.takeBytes();
  if (payload.length > _maximumPayloadBytes) {
    throw const DeviceSigningContractException(
      DeviceSigningErrorCode.payloadTooLarge,
      'The signed payload exceeds one MiB.',
    );
  }
  return payload;
}

List<String> normalizeDeviceScopes(Iterable<String> scopes) {
  final normalized = scopes.toList()..sort();
  if (normalized.isEmpty || normalized.length > deviceScopes.length) {
    throw const DeviceSigningContractException(
      DeviceSigningErrorCode.invalidScope,
      'At least one valid device scope is required.',
    );
  }
  for (var index = 0; index < normalized.length; index += 1) {
    final scope = normalized[index];
    if (!deviceScopes.contains(scope)) {
      throw const DeviceSigningContractException(
        DeviceSigningErrorCode.invalidScope,
        'The requested device scope is not supported.',
      );
    }
    if (index > 0 && normalized[index - 1] == scope) {
      throw const DeviceSigningContractException(
        DeviceSigningErrorCode.duplicateScope,
        'A device scope is duplicated.',
      );
    }
  }
  return List.unmodifiable(normalized);
}

List<SignedPayloadField> _scopeFields(Iterable<String> scopes) {
  final normalized = normalizeDeviceScopes(scopes);
  return [
    for (var index = 0; index < normalized.length; index += 1)
      SignedPayloadField(
        'scope.${index.toString().padLeft(3, '0')}',
        normalized[index],
      ),
  ];
}

String _requireSha256(String value) {
  if (!_sha256Pattern.hasMatch(value)) {
    throw const DeviceSigningContractException(
      DeviceSigningErrorCode.invalidSha256,
      'A signed SHA-256 value is invalid.',
    );
  }
  return value;
}

Uint8List pairingProofPayload({
  required String pairingId,
  required List<int> challenge,
  required String gatewayAudience,
  required String deviceFingerprint,
  required Iterable<String> requestedScopes,
}) {
  return canonicalSignedPayload('agent-talk/pairing-proof/v1', [
    SignedPayloadField('pairing_id', pairingId),
    SignedPayloadField('challenge', challenge),
    SignedPayloadField('gateway_audience', gatewayAudience),
    SignedPayloadField('device_fingerprint', deviceFingerprint),
    ..._scopeFields(requestedScopes),
  ]);
}

Uint8List pairingConfirmationPayload({
  required String pairingId,
  required String credentialId,
  required String deviceId,
  required List<int> challenge,
  required String gatewayAudience,
  required String deviceFingerprint,
  required Iterable<String> approvedScopes,
}) {
  return canonicalSignedPayload('agent-talk/pairing-confirmation/v1', [
    SignedPayloadField('pairing_id', pairingId),
    SignedPayloadField('credential_id', credentialId),
    SignedPayloadField('device_id', deviceId),
    SignedPayloadField('challenge', challenge),
    SignedPayloadField('gateway_audience', gatewayAudience),
    SignedPayloadField('device_fingerprint', deviceFingerprint),
    ..._scopeFields(approvedScopes),
  ]);
}

Uint8List administratorPairingPayload({
  required String pairingId,
  required String userCode,
  required String deviceFingerprint,
  required String gatewayFingerprint,
  required String gatewayAudience,
  required Iterable<String> approvedScopes,
  required List<int> nonce,
}) {
  return canonicalSignedPayload('agent-talk/pairing-approval/v1', [
    SignedPayloadField('pairing_id', pairingId),
    SignedPayloadField('user_code', userCode),
    SignedPayloadField('device_fingerprint', deviceFingerprint),
    SignedPayloadField('gateway_fingerprint', gatewayFingerprint),
    SignedPayloadField('gateway_audience', gatewayAudience),
    SignedPayloadField('nonce', nonce),
    ..._scopeFields(approvedScopes),
  ]);
}

Uint8List credentialRefreshPayload({
  required String credentialId,
  required String deviceId,
  required String gatewayAudience,
  required String refreshTokenSha256,
  required int generation,
  required List<int> nonce,
}) {
  if (generation <= 0) {
    throw const DeviceSigningContractException(
      DeviceSigningErrorCode.invalidField,
      'The credential generation must be positive.',
    );
  }
  return canonicalSignedPayload('agent-talk/credential-refresh/v1', [
    SignedPayloadField('credential_id', credentialId),
    SignedPayloadField('device_id', deviceId),
    SignedPayloadField('gateway_audience', gatewayAudience),
    SignedPayloadField(
      'refresh_token_sha256',
      _requireSha256(refreshTokenSha256),
    ),
    SignedPayloadField('generation', generation.toString()),
    SignedPayloadField('nonce', nonce),
  ]);
}

Uint8List deviceRevocationPayload({
  required String administratorDeviceId,
  required String targetDeviceId,
  required String reasonCode,
  required String gatewayAudience,
  required List<int> nonce,
}) {
  return canonicalSignedPayload('agent-talk/device-revocation/v1', [
    SignedPayloadField('administrator_device_id', administratorDeviceId),
    SignedPayloadField('target_device_id', targetDeviceId),
    SignedPayloadField('reason_code', reasonCode),
    SignedPayloadField('gateway_audience', gatewayAudience),
    SignedPayloadField('nonce', nonce),
  ]);
}

Uint8List approvalDecisionPayload({
  required String credentialId,
  required String deviceId,
  required String hostIdentity,
  required String gatewayAudience,
  required String requestId,
  required String approvalId,
  required String decision,
  required String operationSummarySha256,
  required List<int> nonce,
}) {
  if (decision != 'approve' && decision != 'deny') {
    throw const DeviceSigningContractException(
      DeviceSigningErrorCode.invalidField,
      'The approval decision is invalid.',
    );
  }
  return canonicalSignedPayload('agent-talk/approval-decision/v1', [
    SignedPayloadField('credential_id', credentialId),
    SignedPayloadField('device_id', deviceId),
    SignedPayloadField('host_identity', hostIdentity),
    SignedPayloadField('gateway_audience', gatewayAudience),
    SignedPayloadField('request_id', requestId),
    SignedPayloadField('approval_id', approvalId),
    SignedPayloadField('decision', decision),
    SignedPayloadField(
      'operation_summary_sha256',
      _requireSha256(operationSummarySha256),
    ),
    SignedPayloadField('nonce', nonce),
  ]);
}

Uint8List ownerBootstrapPayload({
  required String gatewayAudience,
  required String deviceFingerprint,
  required Iterable<String> scopes,
  required List<int> nonce,
}) {
  return _ownerPayload(
    domain: 'agent-talk/owner-bootstrap/v1',
    gatewayAudience: gatewayAudience,
    deviceFingerprint: deviceFingerprint,
    scopes: scopes,
    nonce: nonce,
  );
}

Uint8List ownerRecoveryPayload({
  required String gatewayAudience,
  required String deviceFingerprint,
  required Iterable<String> scopes,
  required List<int> nonce,
}) {
  return _ownerPayload(
    domain: 'agent-talk/owner-recovery/v1',
    gatewayAudience: gatewayAudience,
    deviceFingerprint: deviceFingerprint,
    scopes: scopes,
    nonce: nonce,
  );
}

Uint8List _ownerPayload({
  required String domain,
  required String gatewayAudience,
  required String deviceFingerprint,
  required Iterable<String> scopes,
  required List<int> nonce,
}) {
  return canonicalSignedPayload(domain, [
    SignedPayloadField('gateway_audience', gatewayAudience),
    SignedPayloadField('device_fingerprint', deviceFingerprint),
    SignedPayloadField('nonce', nonce),
    ..._scopeFields(scopes),
  ]);
}

File: packages/protocol/src/device-signing.ts
const encoder = new TextEncoder();
const payloadMagic = encoder.encode("agent-talk-signed-payload\0v1");
const domainPattern = /^agent-talk\/[a-z0-9.-]+\/v1$/u;
const fieldNamePattern = /^[a-z][a-z0-9_.-]*$/u;
const sha256Pattern = /^[0-9a-f]{64}$/u;

export const deviceScopes = ["observe", "send", "interrupt", "approve", "administer"] as const;
export type DeviceScope = (typeof deviceScopes)[number];

export type DeviceSigningContractErrorCode =
  | "invalid_domain"
  | "invalid_field"
  | "duplicate_field"
  | "payload_too_large"
  | "invalid_scope"
  | "duplicate_scope"
  | "invalid_sha256";

export class DeviceSigningContractError extends Error {
  constructor(
    readonly code: DeviceSigningContractErrorCode,
    message: string,
  ) {
    super(message);
    this.name = "DeviceSigningContractError";
  }
}

export interface SignedPayloadField {
  name: string;
  value: string | Uint8Array;
}

function bytes(value: string | Uint8Array): Uint8Array {
  return typeof value === "string" ? encoder.encode(value) : value;
}

function uint32(value: number): Uint8Array {
  const encoded = new Uint8Array(4);
  new DataView(encoded.buffer).setUint32(0, value, false);
  return encoded;
}

function frame(value: Uint8Array): Uint8Array[] {
  if (value.byteLength > 1_048_576) {
    throw new DeviceSigningContractError("payload_too_large", "A signed payload field exceeds one MiB.");
  }
  return [uint32(value.byteLength), value];
}

export function canonicalSignedPayload(domain: string, fields: readonly SignedPayloadField[]): Uint8Array {
  if (!domainPattern.test(domain)) {
    throw new DeviceSigningContractError("invalid_domain", "The device-signature domain is invalid.");
  }
  const names = new Set<string>();
  const parts: Uint8Array[] = [payloadMagic, ...frame(encoder.encode(domain)), uint32(fields.length)];
  for (const field of fields) {
    if (!fieldNamePattern.test(field.name)) {
      throw new DeviceSigningContractError("invalid_field", "A device-signature field name is invalid.");
    }
    if (names.has(field.name)) {
      throw new DeviceSigningContractError("duplicate_field", "A device-signature field name is duplicated.");
    }
    names.add(field.name);
    parts.push(...frame(encoder.encode(field.name)), ...frame(bytes(field.value)));
  }
  const length = parts.reduce((total, part) => total + part.byteLength, 0);
  if (length > 1_048_576) {
    throw new DeviceSigningContractError("payload_too_large", "The signed payload exceeds one MiB.");
  }
  const output = new Uint8Array(length);
  let offset = 0;
  for (const part of parts) {
    output.set(part, offset);
    offset += part.byteLength;
  }
  return output;
}

export function normalizeDeviceScopes(scopes: readonly string[]): readonly DeviceScope[] {
  if (scopes.length === 0 || scopes.length > deviceScopes.length) {
    throw new DeviceSigningContractError("invalid_scope", "At least one valid device scope is required.");
  }
  const allowed = new Set<string>(deviceScopes);
  const normalized = [...scopes].sort();
  for (let index = 0; index < normalized.length; index += 1) {
    const scope = normalized[index];
    if (scope === undefined || !allowed.has(scope)) {
      throw new DeviceSigningContractError("invalid_scope", "The requested device scope is not supported.");
    }
    if (index > 0 && normalized[index - 1] === scope) {
      throw new DeviceSigningContractError("duplicate_scope", "A device scope is duplicated.");
    }
  }
  return normalized as DeviceScope[];
}

function scopeFields(scopes: readonly string[]): SignedPayloadField[] {
  return normalizeDeviceScopes(scopes).map((scope, index) => ({
    name: `scope.${index.toString().padStart(3, "0")}`,
    value: scope,
  }));
}

function requireSha256(value: string): string {
  if (!sha256Pattern.test(value)) {
    throw new DeviceSigningContractError("invalid_sha256", "A signed SHA-256 value is invalid.");
  }
  return value;
}

export interface PairingProofPayloadInput {
  pairingId: string;
  challenge: Uint8Array;
  gatewayAudience: string;
  deviceFingerprint: string;
  requestedScopes: readonly string[];
}

export function pairingProofPayload(input: PairingProofPayloadInput): Uint8Array {
  return canonicalSignedPayload("agent-talk/pairing-proof/v1", [
    { name: "pairing_id", value: input.pairingId },
    { name: "challenge", value: input.challenge },
    { name: "gateway_audience", value: input.gatewayAudience },
    { name: "device_fingerprint", value: input.deviceFingerprint },
    ...scopeFields(input.requestedScopes),
  ]);
}

export interface PairingConfirmationPayloadInput {
  pairingId: string;
  credentialId: string;
  deviceId: string;
  challenge: Uint8Array;
  gatewayAudience: string;
  deviceFingerprint: string;
  approvedScopes: readonly string[];
}

export function pairingConfirmationPayload(input: PairingConfirmationPayloadInput): Uint8Array {
  return canonicalSignedPayload("agent-talk/pairing-confirmation/v1", [
    { name: "pairing_id", value: input.pairingId },
    { name: "credential_id", value: input.credentialId },
    { name: "device_id", value: input.deviceId },
    { name: "challenge", value: input.challenge },
    { name: "gateway_audience", value: input.gatewayAudience },
    { name: "device_fingerprint", value: input.deviceFingerprint },
    ...scopeFields(input.approvedScopes),
  ]);
}

export interface AdministratorPairingPayloadInput {
  pairingId: string;
  userCode: string;
  deviceFingerprint: string;
  gatewayFingerprint: string;
  gatewayAudience: string;
  approvedScopes: readonly string[];
  nonce: Uint8Array;
}

export function administratorPairingPayload(input: AdministratorPairingPayloadInput): Uint8Array {
  return canonicalSignedPayload("agent-talk/pairing-approval/v1", [
    { name: "pairing_id", value: input.pairingId },
    { name: "user_code", value: input.userCode },
    { name: "device_fingerprint", value: input.deviceFingerprint },
    { name: "gateway_fingerprint", value: input.gatewayFingerprint },
    { name: "gateway_audience", value: input.gatewayAudience },
    { name: "nonce", value: input.nonce },
    ...scopeFields(input.approvedScopes),
  ]);
}

export interface CredentialRefreshPayloadInput {
  credentialId: string;
  deviceId: string;
  gatewayAudience: string;
  refreshTokenSha256: string;
  generation: bigint;
  nonce: Uint8Array;
}

export function credentialRefreshPayload(input: CredentialRefreshPayloadInput): Uint8Array {
  if (input.generation <= 0n) {
    throw new DeviceSigningContractError("invalid_field", "The credential generation must be positive.");
  }
  return canonicalSignedPayload("agent-talk/credential-refresh/v1", [
    { name: "credential_id", value: input.credentialId },
    { name: "device_id", value: input.deviceId },
    { name: "gateway_audience", value: input.gatewayAudience },
    { name: "refresh_token_sha256", value: requireSha256(input.refreshTokenSha256) },
    { name: "generation", value: input.generation.toString() },
    { name: "nonce", value: input.nonce },
  ]);
}

export interface DeviceRevocationPayloadInput {
  administratorDeviceId: string;
  targetDeviceId: string;
  reasonCode: string;
  gatewayAudience: string;
  nonce: Uint8Array;
}

export function deviceRevocationPayload(input: DeviceRevocationPayloadInput): Uint8Array {
  return canonicalSignedPayload("agent-talk/device-revocation/v1", [
    { name: "administrator_device_id", value: input.administratorDeviceId },
    { name: "target_device_id", value: input.targetDeviceId },
    { name: "reason_code", value: input.reasonCode },
    { name: "gateway_audience", value: input.gatewayAudience },
    { name: "nonce", value: input.nonce },
  ]);
}

export interface ApprovalDecisionPayloadInput {
  credentialId: string;
  deviceId: string;
  hostIdentity: string;
  gatewayAudience: string;
  requestId: string;
  approvalId: string;
  decision: "approve" | "deny";
  operationSummarySha256: string;
  nonce: Uint8Array;
}

export function approvalDecisionPayload(input: ApprovalDecisionPayloadInput): Uint8Array {
  return canonicalSignedPayload("agent-talk/approval-decision/v1", [
    { name: "credential_id", value: input.credentialId },
    { name: "device_id", value: input.deviceId },
    { name: "host_identity", value: input.hostIdentity },
    { name: "gateway_audience", value: input.gatewayAudience },
    { name: "request_id", value: input.requestId },
    { name: "approval_id", value: input.approvalId },
    { name: "decision", value: input.decision },
    { name: "operation_summary_sha256", value: requireSha256(input.operationSummarySha256) },
    { name: "nonce", value: input.nonce },
  ]);
}

export interface OwnerBootstrapPayloadInput {
  gatewayAudience: string;
  deviceFingerprint: string;
  scopes: readonly string[];
  nonce: Uint8Array;
}

export function ownerBootstrapPayload(input: OwnerBootstrapPayloadInput): Uint8Array {
  return canonicalSignedPayload("agent-talk/owner-bootstrap/v1", [
    { name: "gateway_audience", value: input.gatewayAudience },
    { name: "device_fingerprint", value: input.deviceFingerprint },
    { name: "nonce", value: input.nonce },
    ...scopeFields(input.scopes),
  ]);
}
